src/pykiso/lib/connectors/cc_process.py

```diff
@@ -0,0 +1,319 @@
+#
+# SPDX-License-Identifier: EPL-2.0
+##########################################################################
+
+"""
+Process Channel
+***************
+
+:module: cc_process
+
+:synopsis: CChannel implementation for process execution.
+
+The CCProcess channel provides functionality to start a process and
+to communicate with it.
+
+.. currentmodule:: cc_process
+
+"""
+
+import logging
+import queue
+import subprocess
+import threading
+from dataclasses import dataclass
+from typing import IO, ByteString, Callable, Dict, List, Optional, Tuple, Union
+
+from pykiso import Message
+from pykiso.connector import CChannel
+
+log = logging.getLogger(__name__)
+
+MessageType = Union[str, ByteString]
+
+
+class CCProcessError(BaseException):
+    ...
+
+
+@dataclass
+class ProcessMessage:
+    """Holds the data that is read from the process"""
+
+    # Stream name: stdout or stderr
+    stream: str
+    # Data as bytes or string
+    data: Union[str, bytes]
+
+
+class ProcessExit:
+    ...
+
+
+# Marker for process exit in queue_in
+PROCESS_EXIT = ProcessExit()
+
+
+class CCProcess(CChannel):
+    """Channel to run processes"""
+
+    def __init__(
+        self,
+        shell: bool = False,
+        pipe_stderr: bool = False,
+        pipe_stdout: bool = True,
+        pipe_stdin: bool = False,
+        text: bool = True,
+        cwd: Optional[str] = None,
+        env: Optional[str] = None,
+        encoding: Optional[str] = None,
+        executable: Optional[str] = None,
+        args: List[str] = [],
+        **kwargs,
+    ):
+        """Initialize a process
+
+        :param shell: Start process through shell
+        :param pipe_stderr: Pipe stderr for reading with this connector
+        :param pipe_stdout: Pipe stdout for reading with this connector
+        :param pipe_stdin:  Pipe stdin for writing with this connector
+        :param text: Read/write stdout, stdin, stderr in binary mode
+        :param cwd: The current working directory for the new process
+        :param env: Environment variables for the new process
+        :param encoding: Encoding to use in text mode
+        :param executable: The path of the executable for the process
+        :param args: Process arguments
+
+        """
+        super().__init__(**kwargs)
+        self.shell = shell
+        self.pipe_stderr = pipe_stderr
+        self.pipe_stdout = pipe_stdout
+        self.pipe_stdin = pipe_stdin
+        self.encoding = encoding
+        self.executable = executable
+        self.args = args
+        self.text = text
+        self.cwd = cwd
+        self.env = env
+        self.process: Optional[subprocess.Popen] = None
+        self.queue_in: Optional[queue.Queue[Union[ProcessMessage, ProcessExit]]] = None
+        self.stdout_thread: Optional[threading.Thread] = None
+        self.stderr_thread: Optional[threading.Thread] = None
+        self.lock: threading.Lock = threading.Lock()
+        # Counter for finished threads
+        self.ready: int = 0
```

Comment:
Also, if you'd use a more explicit name for the variable you wouldn't need to add a comment to explain its goal 

And IMO this should be private

Reply:
Done

Reply:
Done